File: academic-system/src/main/java/org/example/academic/system/AcademicSystem.java
package org.example.academic.system;

import org.example.academic.system.model.*;

import java.util.ArrayList;
import java.util.List;

public class AcademicSystem {

    private static AcademicSystem instance;
    private List<AcademicClass> classes;

    private AcademicSystem() {
        classes = new ArrayList<>();
    }

    public static AcademicSystem getInstance() {
        if (instance == null) {
            instance = new AcademicSystem();
        }
        return instance;
    }

    public void addClass(AcademicClass academicClass) {
        classes.add(academicClass);
    }

    public List<AcademicClass> getClasses() {
        return classes;
    }

    public AcademicClass findClassByCode(String code) {
        for (AcademicClass academicClass : classes) {
            if (academicClass.getCode().equals(code)) {
                return academicClass;
            }
        }
        return null;
    }
}

File: academic-system-python/src/academic_system/academic_system.py
from __future__ import annotations
from typing import List, Optional
from academic_system.model.academic_class import AcademicClass


class AcademicSystem:
    _instance: Optional[AcademicSystem] = None

    def __init__(self):
        self._classes: List[AcademicClass] = []

    @classmethod
    def get_instance(cls) -> AcademicSystem:
        if cls._instance is None:
            cls._instance = AcademicSystem()
        return cls._instance

    @classmethod
    def reset_instance(cls):
        """Usado apenas nos testes."""
        cls._instance = None

    def add_class(self, academic_class: AcademicClass):
        self._classes.append(academic_class)

    def get_classes(self) -> List[AcademicClass]:
        return self._classes

    def find_class_by_code(self, code: str) -> Optional[AcademicClass]:
        for c in self._classes:
            if c.code == code:
                return c
        return None